examples/minimal-mdns/server.cpp

```diff
@@ -83,31 +96,55 @@ OptionSet cmdLineOptions = { HandleOptions, cmdLineOptionsDef, "PROGRAM OPTIONS"
                              "  -4\n"
                              "  --enable-ip-v4\n"
                              "        enable listening on IPv4\n"
+                             "  -i <name>\n"
+                             "  --instance-name <name>\n"
+                             "        instance name to advertise.\n"
                              "\n" };
 
 HelpOptions helpOptions("minimal-mdns-server", "Usage: minimal-mdns-server [options]", "1.0");
 
 OptionSet * allOptions[] = { &cmdLineOptions, &helpOptions, nullptr };
 
-class ReplyDelegate : public mdns::Minimal::ServerDelegate
+class ReplyDelegate : public mdns::Minimal::ServerDelegate, public mdns::Minimal::ParserDelegate
 {
 public:
+    ReplyDelegate(mdns::Minimal::ResponseSender * responder) : mResponder(responder) {}
+
     void OnQuery(const mdns::Minimal::BytesRange & data, const chip::Inet::IPPacketInfo * info) override
     {
-        char addr[32];
+        char addr[INET6_ADDRSTRLEN];
         info->SrcAddress.ToString(addr, sizeof(addr));
 
         printf("QUERY from: %-15s on port %d, via interface %d\n", addr, info->SrcPort, info->Interface);
         Report("QUERY: ", data);
+
+        mCurrentSource = info;
+        if (!mdns::Minimal::ParsePacket(data, this))
+        {
+            printf("Parsing failure may result in reply failure!\n");
+        }
+        mCurrentSource = nullptr;
     }
 
     void OnResponse(const mdns::Minimal::BytesRange & data, const chip::Inet::IPPacketInfo * info) override
     {
         char addr[32];
         info->SrcAddress.ToString(addr, sizeof(addr));
 
-        printf("RESPONSE from: %-15s on port %d, via interface %d\n", addr, info->SrcPort, info->Interface);
-        Report("RESPONSE: ", data);
+        // printf("RESPONSE from: %-15s on port %d, via interface %d\n", addr, info->SrcPort, info->Interface);
+        // Report("RESPONSE: ", data);
+    }
+
+    // ParserDelegate
+    void OnHeader(mdns::Minimal::ConstHeaderRef & header) override { mMessageId = header.GetMessageId(); }
```

Comment:
For a client listening to multiple mdns services (e.g. `_chipc._udp` and _chip._tcp`), the current api of `ParserDelegate` makes it hard to tell what's the packet we are handing, can we also pass the full packet or some identifier to the delegate as well?

Reply:
I was thinking to use the transaction id in the header if we need multiple listening. Otherwise we would need to check the query section of a parsing.

At this point of the server I only have the header information (I just received a packet and started parsin it) so the info here is all that is available. Nothing more can be added.